### PR TITLE
make home path configurable ... continuation of #76114

### LINKED ISSUE #75788: Ability to easy control of ~/.ansible directory location
### Summary

I want to control ~/.ansible directory location but it's ridiculously complex. My main motivation is clean and controllable $HOME directory.

Currently I have to setup around 24 environment variables to make it possible. And even if achieve it, it is not guaranteed that behavior will not be changed in the future or new plugin will not require to fix the path.

I'm aware of previous attempts to adopt XDG:
- https://github.com/ansible/ansible/issues/52354
- https://github.com/ansible/ansible/issues/68587

I'm partially agree your points against XDG. For me personally it's not XDG_CONFIG, it should be XDG_DATA... or XDG_CACHE.

I suggest to create **one** variable which will be base prefix of rest ~**24** variables which refers to ~/.ansible directory. This variable may not have any relations with XDG at all but allow to get control to users and get rid of polluted $HOMEs.

```yaml
BASE_PREFIX:
  default: ~/.ansible
COLLECTIONS_PATHS:
  default: {{BASE_PREFIX}}/collections:/usr/share/ansible/collections
```

I offer my help to work on PR as well if you will not be against of this idea.

### Issue Type

Feature Idea

### Component Name

core

### Additional Information

```yaml (paste below)

```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by stefanwascoding**
##### SUMMARY
This is a continuation of the work done in #76114 to get an ANSIBLE_HOME configuration variable working.

Fixes #75788

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
base.yml

### CODE REVIEW

**Comment by stefanwascoding** [comment]
starting with a rebase to latest devel, to see if this changes the tests.

**Comment by stefanwascoding** [comment]
Well, the tests are looking a lot better with that last WIP change ... now just to find out why the path seems to get messed up in the galaxy unit tests ( [example error in recent run](https://dev.azure.com/ansible/ansible/_build/results?buildId=38374&view=logs&j=40bacf7d-165d-5b57-35fc-8a6ffc0ddfca&t=92cdabea-8088-50d2-348b-96ccb91c1989&l=33818) -- failed path "unfrack"-ing‽ ), and why the collections path does errors ( [working run for someone else](https://dev.azure.com/ansible/ansible/_build/results?buildId=38350&view=logs&j=05630f17-7677-51af-4b88-11118adc6a44&t=b3edd11b-a13a-5494-96a8-2c9d98447f6f&l=2003) VS [error run on our branch](https://dev.azure.com/ansible/ansible/_build/results?buildId=38373&view=logs&j=05630f17-7677-51af-4b88-11118adc6a44&t=b3edd11b-a13a-5494-96a8-2c9d98447f6f&l=2017) )

**Comment by stefanwascoding** [comment]
Question for the maintainers ... https://github.com/ansible/ansible/commit/6219fc4932797f29d23d331bbdaa6b3eed24f230 added an `ANSIBLE_HOME` environment variable to the `hacking/env-setup` script.  I feel like either the new config var, or that one should be renamed, so as not to step on each other's toes.
If you agree, any preference which one?
From what I can tell by `grep`-ing through the source, the env var from the hacking script was not used outside that script (at least in the ansible repo).

**Comment by bcoca** [comment]
my 2c modify the hacking script to a more obscure name (ANSIBLE_DEV_HOME?) and keep the user friendly one

File: lib/ansible/cli/galaxy.py [hunk 1]
```diff
@@ -1508,7 +1508,7 @@ def execute_list_collection(self, artifacts_manager=None):
         """
 
         output_format = context.CLIARGS['output_format']
-        collections_search_paths = set(context.CLIARGS['collections_path'])
+        collections_search_paths = set([C.COLLECTIONS_PATHS[0]])
```
Comment (sivel) [inline lib/ansible/cli/galaxy.py]: I don't think this is right. You are removing the ability for the collection path to be set on the CLI, in favor of only accepting the configuration value.

Comment (stefanwascoding) [reply lib/ansible/cli/galaxy.py]: You are right, and it did bring new errors  (but different ones!) ... this change does seem to at least fix the error about not finding the (default) collection path [see here](https://dev.azure.com/ansible/ansible/_build/results?buildId=38621&view=logs&j=05630f17-7677-51af-4b88-11118adc6a44&t=b3edd11b-a13a-5494-96a8-2c9d98447f6f&l=2505).
So now I need to find a way to have the `context.CLIARGS` handle default parameters that contain template values.

Comment (sivel) [reply lib/ansible/cli/galaxy.py]: Yesterday I responded in IRC, that the problem was due to using `C.config.get_config_value('CONFIG_NAME)` elsewhere in the codebase, and that they would need to be switched to use `C.CONFIG_NAME`.  Only uses of `C.CONFIG_NAME` have templating applied to them, and use of `C.config.get_config_value()` does not.

This is the specific line I am referring to:

https://github.com/ansible/ansible/blob/7c81a652c0af1831bb7e0afd0bb23c43e68a40bf/lib/ansible/plugins/loader.py#L1148

Comment (stefanwascoding) [reply lib/ansible/cli/galaxy.py]: That worked perfectly.

Except for the slight wrinkle where my (overeager?) change of a similar construct here: https://github.com/ansible/ansible/pull/77322/commits/a40927794960fa02dcd26a01010d74f816a30108#diff-fe8d3e13b2884db3e502001befe5ca81b0b203fe2cec9b31e2ee9b0f14a8e835R273 causes errors in some galaxy cache handling unit tests.

As an example, constant access returns: `b'/root/.ansible/galaxy_cache'` whereas the `get_config_value()` returns: `
b'/tmp/pytest-of-root/pytest-7/popen-gw0/Test \xc3\x85\xc3\x91\xc5\x9a\xc3\x8c\xce\xb2\xc5\x81\xc3\x88 Galaxy Cache0'`

**Comment by stefanwascoding** [comment]
When `C.config.get_config_value("GALAXY_CACHE_DIR")` is used in https://github.com/ansible/ansible/blob/babc26adc1ee0ac054d9a0ab19784988cef867b4/lib/ansible/galaxy/api.py#L273 the existing [galaxy api unit tests](https://github.com/ansible/ansible/blob/devel/test/units/galaxy/test_api.py) pass, but the cache folder is created in the wrong location when using `$ANSIBLE_HOME` (no templating is applied, so you get a folder structure on disk that is actually named `./{{ ANSIBLE_HOME ~ " /galaxy_cache" }}` ).

When `C.GALAXY_CACHE_DIR` is used, you get template handling and the expected paths on disk, but some galaxy api unit tests fail (as in these [CI logs](https://dev.azure.com/ansible/ansible/_build/results?buildId=38659&view=logs&j=570d6be3-a35a-586c-0b02-6634bbb9f96c&t=364ad5e9-b72d-51c6-d820-e6e985ea9085))

Is there a way to access config settings, that supports both dynamic changes (as `get_config_value()` does, since the tests do seem to make use of some feature there) as well as templating (like the constant access). 
How to get the best of both worlds - template handling & passing unit tests?

There was a [comment on Matrix](https://matrix.to/#/!DyJEpptWlrIexlAmAY:libera.chat/$xwYVx3VvMrTKL5mvEtA21ar4TuRWT_N9dF76w5OC5S0?via=ansible.com&via=libera.chat&via=matrix.org) by @sivel about monkeypatching "`C` instead of `C.config._base_defs`". If this is already a proposed solution to this problem, then TBH I do not understand what is meant by that & what that needs to be done, and could someone please clarify.

**Comment by sivel** [comment]
> Is there a way to access config settings, that supports both dynamic changes (as `get_config_value()` does, since the tests do seem to make use of some feature there) as well as templating (like the constant access).

You would have to update `get_config_value` to do what is done in:

https://github.com/ansible/ansible/blob/394d216922d70709248a60f58da300f1e70f5894/lib/ansible/constants.py#L183-L202

That implementation isn't perfect either, as it seems to depend on the order of config keys.

**Comment by stefanwascoding** [comment]
> You would have to update `get_config_value` to do what is done in:

I think this is better off done in a separate PR.

**Comment by stefanwascoding** [comment]
latest force push to clean up history: get rid of intermediate & WIP commits, remove debug code, ...

**Comment by stefanwascoding** [comment]
Closing PR in favor of collaborating on #76114.

**Pull request closed without merging**